[FEATURE] 캘린더 월별 지원서 목록 조회 API 생성

Seongwon97 opened the pull request:
## 작업 내용
- 캘린더 월별 지원서 목록 조회 api를 생성했습니다.

## 논의 사항
- 응답 값에서 data 내부에 applications으로 감싸고 안에 List형태로 반환해줄지 data안에 바로 List를 반환할지 고민입니다.
- TestFixture를 원래 domain모듈에 위치시키고 다른 모듈들에서 각각 필요한 형태로 변환한 후에 사용하려했는데...모듈간 implementation을 하여도 main패키지만 의존이 넘어가고 test패키지는 넘어가지 않는다네요ㅠㅠ 이렇게 되면 TestFixture를 모듈간 따로 관리해야하는데..이것 또한 너무 많은 비용이 들 것 같아 고민입니다.

resolve #48

Seongwon97 committed:
feat: 캘린더 월별 조회 in port 정의

Seongwon97 committed:
feat: 캘린더 월별 조회 out port 정의

Seongwon97 committed:
refactor: Process 타입을 record에서 class로 변경

Seongwon97 committed:
feat: 지원서 월별 조회 서비스 로직 구현

Seongwon97 committed:
feat: 조회를 위한 데이터로 연도 데이터 추가

Seongwon97 committed:
build: test 코드에 Lombok 의존성 추가

Seongwon97 committed:
test: ApplicationFixture생성

Seongwon97 committed:
test: ProcessFixture생성

Seongwon97 committed:
test: persistence 테스트 환경 설정

Seongwon97 committed:
feat: userId와 일정을 통해 이력서 조회를 하는 JPQL 쿼리 생성

Seongwon97 committed:
refactor: 지원서 도메인의 deadline오타 수정

Seongwon97 committed:
feat: ApplicationEntity의 도메인 변환 메서드 생성

Seongwon97 committed:
test: ApplicationFixture의 엔티티 생성 로직에 immutableList가 아닌 mutableList를 사용하도록 변경

Seongwon97 committed:
feat: 사용자와 날짜 정보를 통해 지원서 정보를 읽어오는 Adapter구현

Seongwon97 committed:
feat: ApplicationQueryService에 Transactional 어노테이션 추가

Seongwon97 committed:
feat: 지원서 월간 데이터 조회 API 생성

Seongwon97 committed:
refactor: url을 복수형으로 수정

Seongwon97 committed:
build: application모듈에 spring-starter-test의존성 추가

Seongwon97 committed:
test: 테스트 네이밍 수정

Seongwon97 committed:
test: 누락된 ApplicationQueryService 테스크 추가

Seongwon97 commented on out-adapter-persistence-jpa/src/test/resources/application.yml:
```diff
@@ -0,0 +1,4 @@
+spring:
+  config:
+    import:
+      - application-persistence-local.yml
```
해당 파일은 유빈님 PR이 먼저 머지되면 file change에서 없어질 코드입니다!

Seongwon97 commented on core-application/src/test/java/gohigher/application/service/ApplicationQueryServiceTest.java:
```diff
@@ -0,0 +1,85 @@
+		private final int year = 2023;
+		private final int month = 9;
+
+		@DisplayName("해당 연/월에 유저가 등록한 지원서 일정이 정보가 있을 때")
+		@Nested
+		class Context_with_schedules {
+
+			private final LocalDate searchDate = LocalDate.of(year, month, 20);
+			private final Process toApply = new Process(ProcessType.TO_APPLY, null,
+				LocalDateTime.of(searchDate, LocalTime.now()));
+			private final Process document = new Process(ProcessType.DOCUMENT, null,
+				LocalDateTime.of(searchDate, LocalTime.now()));
+			private final Process interview = new Process(ProcessType.INTERVIEW, null,
+				LocalDateTime.of(searchDate, LocalTime.now()));
+
+			@BeforeEach
+			void setUp() {
+				List<Process> naverProcesses = List.of(toApply, document, interview);
+				Application naverApplication = new Application("Naver", "Bakcend", "", "", "", "",
+					EmploymentType.PERMANENT, "", "", "", LocalDateTime.now().plusDays(7), naverProcesses, "", toApply);
+
+				List<Process> kakaoProcesses = List.of(toApply, document);
+				Application kakaoApplication = new Application("Kakao", "Bakcend", "", "", "", "",
+					EmploymentType.PERMANENT, "", "", "", LocalDateTime.now().plusDays(7), kakaoProcesses, "", toApply);
+
+				List<Application> applicationJpaEntities = List.of(naverApplication, kakaoApplication);
+
+				given(applicationPersistenceQueryPort.findByIdAndMonth(userId, year, month))
+					.willReturn(applicationJpaEntities);
+			}
+
+			@DisplayName("일정 정보를 반환한다.")
+			@Test
+			void it_return_application_processes() {
+				ApplicationMonthQueryResponse actual = applicationQueryService.findByMonth(userId, year, month);
+
+				assertThat(actual.getApplications()).hasSize(5);
+			}
+		}
```
[유빈님의 테스트 관련 PR에서 나누었던 이야기](https://github.com/team-go-higher/backend/pull/47#discussion_r1317127924)에서 제가 생각한 테스트 코드는 다음과 같이 최하단 클래스는 Given의 느낌을 주는 클래스라, 클래스 내에서 given의 설정을 하고 메서드 레벨에서 세부적인 설정과 when,then을 체크하는 방법을 생각했었습니다!

Seongwon97 committed:
style: 누락된 EOF 공백라인 추가

kyukong commented on core-application/src/main/java/gohigher/application/port/in/ApplicationCalendarResponse.java:
```diff
@@ -0,0 +1,15 @@
+
+@Getter
+@RequiredArgsConstructor
+public class ApplicationCalendarResponse {
```
DTO 이름을 CalendarApplication~~ 으로 바꾸는 것은 어떨까요? 약간 어색하긴 하지만 Calendar~~ 로 시작하면 유사한 DTO 끼리 묶여보여서 좋더라구요. 앞으로 Application 관해서 DTO 가 매우 많아질것 같은데 이름으로 유사 DTO 를 군집화하는건 어떨까요!

Seongwon97 replied:
좋은 생각입니다!! `CalenderApplicationResponse`으로 변경했습니다!

kyukong commented on core-application/src/main/java/gohigher/application/port/in/ApplicationMonthQueryResponse.java:
```diff
@@ -0,0 +1,13 @@
+
+@Getter
+@RequiredArgsConstructor
+public class ApplicationMonthQueryResponse {
```
DTO 이름에 Query 는 어떤 의미인가요?

Seongwon97 replied:
월별 조회라는 의미를 주려고 붙이긴 했었는데 다시 보니까 이상한 것 같네요..!

위의 클래스와 함께 `CalenderApplicationMonthResponse`로 변경했습니다!

kyukong commented on core-domain/src/main/java/gohigher/common/JobInfo.java:
```diff
@@ -18,7 +17,7 @@
  * requiredCapability     필수 역량
  * preferredQualification 우대사항
- * deadLine               마감 날짜(상시면 해당 년의 마지막 날로 설정)
+ * deadline               마감 날짜(상시면 해당 년의 마지막 날로 설정)
```
👍

kyukong commented on out-adapter-persistence-jpa/src/main/java/gohigher/application/entity/ApplicationRepository.java:
```diff
@@ -1,6 +1,16 @@
+		+ "WHERE a.userId = :userId "
+		+ "AND FUNCTION('YEAR', p.schedule) = :year "
+		+ "AND FUNCTION('MONTH', p.schedule) = :month")
+	List<ApplicationJpaEntity> findByUserIdAndMonth(Long userId, int year, int month);
```
메서드 네임으로는 user id 와 month 만 받을것 같은데 year 정보도 받네요! `AndYearAndMonth` 나 `AndDate`, `AndSchedule` 등으로 변경하는건 어떨까요?

Seongwon97 replied:
좋은 의견입니다!! 스케줄은 일정 정보를 받아하는 느낌이 들어서 `findByUserIdAndDate`으로 이름을 수정했습니다!

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/ApplicationPersistenceQueryAdapterTest.java:
```diff
@@ -0,0 +1,81 @@
+					month);
+
+				assertAll(
+					() -> assertThat(applications).hasSize(2),
```
```suggestion
					() -> assertThat(applications).hasSize(applicationJpaEntities.size()),
```
저장 전 리스트의 길이로 받으면 테스트 코드 수정에도 용이하고 2의 의미도 명확히 드러낼 수 있을것 같습니다!

kyukong replied:
그리고 다음에 테스트가 추가될 것을 고려해서 테스트 격리가 필요해보입니다!

Seongwon97 replied:
피드백을 반영하여 `2`와 같은 매직넘버들은 의미를 알 수 있게 수정했습니다!!
의미를 표현하기 힘든 경우는 주석을 추가했습니다!

Seongwon97 replied:
혹시 테스트 격리에 대해서는 어떤 것을 말씀하시는건지 여쭈어봐도 될까요...?
`RepositoryTest`의 경우 `@DataJpaTest`가 `@Transactional` 어노테이션을 갖고 있어서 테스트별로 실행될 때, 데이터베이스에 실제 저장되지 않고 Rollback 처리가 되어 격리가 되는 상태라 생각됩니다. 혹시 우려되는 상황이 있을까요?

kyukong replied:
테스트 시 실제 데이터베이스를 사용하기 때문에, 로컬 데이터베이스에 데이터가 있는 상태에서 메서드를 호출하면 결과와 다르게 나올수도 있을것 같습니다! (저장되어 있던 데이터 + 테스트용으로 저장한 데이터)

Seongwon97 replied:
간단하게 생각한 해결책으로는 현재 문제가 발생할 것이 기존에 저장된 데이터가 테스트에 끼칠 영향을 없애는 것이기에 `@GeneratedValue`에서 사용하지 않는 id값(0, -1 등)을 통해 테스트를 하는 것이 떠올라 적용하였습니다.

kyukong replied:
넵넵 TestEntityManager 로 테스트 격리가 가능하다면 사실 문제 없을거같긴 합니다! ㅎㅎ 직접 테스트 하고 싶었으나 제 코드와 충돌이 나서 확인이 어렵네요. 😅

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/entity/ApplicationRepositoryTest.java:
```diff
@@ -0,0 +1,137 @@
+class ApplicationRepositoryTest {
+
+	@Autowired
+	ApplicationRepository applicationRepository;
```
```suggestion
	private ApplicationRepository applicationRepository;
```
테스트라서 큰 의미는 없지만... 다른 코드와 형식을 맞춰보면 어떨까요!

Seongwon97 replied:
테스트여도 중요하죠..!!
해당 클래스에도 접근제어자를 추가했습니다!

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/entity/ApplicationRepositoryTest.java:
```diff
@@ -0,0 +1,137 @@
+	ApplicationProcessRepository applicationProcessRepository;
+
+	@Autowired
+	EntityManager entityManager;
```
`EntityManager` 대신에 `TestEntityManager` 를 사용하는건 어떤가요? `EntityManager` 는 실제 데이터베이스 연결을 필요로 하지만 `TestEntityManager` 는 Mock 데이터베이스를 사용한다는 장점이 있습니다.

Seongwon97 replied:
`TestEntityManager`는 처음 알았네요..!! 변경 완료했습니다!

kyukong commented on in-adapter-api/src/main/java/gohigher/application/ApplicationQueryController.java:
```diff
@@ -0,0 +1,26 @@
+
+	@GetMapping("/v1/applications/calender")
+	public ResponseEntity<GohigherResponse<ApplicationMonthQueryResponse>> findByMonth(@Login Long userId,
+		@RequestParam int year, @RequestParam int month) {
```
이건 그냥 궁금한 점인데 왜 날짜 입력을 year 와 month 로 나누어서 받나요??

Seongwon97 replied:
해당 부분은 살짝 date로 하고 `yyyy-MM`형식으로 받을지 분리하여 받을지 고민이었습니다... 

queryParam으로 받을 때, `/v1/applications/calender?year=2023&month=10`과 `/v1/applications/calender?month=2023-10`의 차이인데 저는 `2023-10`보다는 쪼개서 받는 것이 좋아보이지 않을까..?하는 생각에 나누어 받았는데 두번째 api의 형태가 더 자연스러워 보이신다면 변경해도 괜찮습니다!

kyukong replied:
그냥 궁금해서 남겨봤습니다! ㅎㅎ 하나로 묶어 전달하는 방식에 더 익숙해져 있다보니 어색해서 여쭈어봤습니다! 나누어서 값을 가져온다면 프론트 쪽에서 날짜 형식에 대해 신경쓸 필요가 없고, 필수 값들을 쉽게 강제할 수 있어서 좋은것 같습니다!! 👍

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/fixture/ApplicationFixture.java:
```diff
@@ -0,0 +1,47 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum ApplicationFixture {
```
enum 굳굳~~~!!! 👍👍👍

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/fixture/ProcessFixture.java:
```diff
@@ -0,0 +1,43 @@
+@RequiredArgsConstructor
+public enum ProcessFixture {
+
+	TO_APPLY(ProcessType.TO_APPLY, "", LocalDateTime.now()),
+	DOCUMENT(ProcessType.DOCUMENT, "", LocalDateTime.now()),
+	TEST(ProcessType.TEST, "", LocalDateTime.now()),
+	INTERVIEW(ProcessType.INTERVIEW, "", LocalDateTime.now()),
+	COMPLETE(ProcessType.COMPLETE, "", LocalDateTime.now()),
+	;
```
테스트 데이터에 LocalDateTime.now() 는 시간 관련해서 테스트 시 어려울듯합니다!

Seongwon97 replied:
기존에는 도메인으로의 변환 메서드에 무조건 일정을 받도록 설정하여 `LocalDateTime.now()`의 의미가 크게 없었는데, 편의를 위해 아래와 같이 임의로 일정을 설정하고 일정 정보를 받지 않고서도 도메인으로 변경하는 로직을 추가했습니다.

```java
	TO_APPLY(ProcessType.TO_APPLY, "", null),
	DOCUMENT(ProcessType.DOCUMENT, "", LocalDateTime.now().plusDays(6)),
	TEST(ProcessType.TEST, "", LocalDateTime.now().plusDays(10)),
	INTERVIEW(ProcessType.INTERVIEW, "", LocalDateTime.now().plusDays(20)),
	COMPLETE(ProcessType.COMPLETE, "", LocalDateTime.now().plusDays(40)),
```

kyukong replied:
으음.. 사실 시간관련 테스트 시에도 사용하고자 남긴 리뷰였는데, 생각해보니 시간 관련 픽스쳐를 따로 만드는 것이 나을것 같다는.. 생각도 드네요..... 😅😅 일단 현상태로 유지해도 좋을것같아요!!

kyukong reviewed (changes requested):
픽스쳐.. 만들때마다 머리아픈데 정말 고생 많으셨습니다 👍
잘 쓰겠습니다..

Seongwon97 committed:
build: java-test-fixture 플러그인 추가

Seongwon97 committed:
test: fixture를 domain 모듈로 이동

Seongwon97 committed:
test: domain 모듈의 fixture 사용을 위한 persistence모듈의 converter생성

Seongwon97 committed:
feat: 연관관계 편의 메서드 생성

Seongwon97 committed:
test: domain 모듈의 테스트 fixture를 사용하도록 persistence 테스트 수정

Seongwon97 committed:
test: service 모듈 테스트에 fixture적용

Seongwon97 committed:
test: 누락된 접근 제어자 추가

Seongwon97 committed:
refactor: 캘린더 조회관련 DTO 네이밍 수정

- ApplicationMonthQueryResponse -> CalenderApplicationMonthResponse
- ApplicationCalendarResponse -> CalenderApplicationResponse

Seongwon97 committed:
refactor: 월별 일정 조회 쿼리 메서드 네이밍 수정

- findByUserIdAndMonth -> findByUserIdAndDate

Seongwon97 committed:
test: 테스트 검증로직 내의 매직넘버 제거

Seongwon97 committed:
test: RepositoryTest의 EntityManager객체를 TestEntityManager로 변경

Seongwon97 committed:
test: processFixture의 일정을 다양하게 설정

Byeongju-Kong commented on core-application/src/test/java/gohigher/application/service/ApplicationQueryServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+
+	private ApplicationQueryService applicationQueryService;
+
+	@BeforeEach
+	void setUp() {
+		applicationQueryService = new ApplicationQueryService(applicationPersistenceQueryPort);
+	}
```
객체를 테스트 마다 생성하시는 이유가 있나요? 해당 객체는 상태 값을 가지고 있지 않기 때문에 1번의 생성으로 충분하다는 생각이 듭니다!

Seongwon97 replied:
`@BeforeEach` 문에서 초기화를 하지 않고 필드 레벨에서 생성을 한다면 아래와 같이 mock 인스턴스가 만들어지기 전이라 NPE가 터지게 됩니다..! 
그리하여 BeforeEach로 매번 넣어주게 하였습니다😭

<img width="887" alt="image" src="https://github.com/team-go-higher/backend/assets/57744251/66f6d9ca-717d-45e9-9a51-21dcc8e22786">

Byeongju-Kong commented on in-adapter-api/src/main/java/gohigher/application/ApplicationQueryControllerDocs.java:
```diff
@@ -0,0 +1,25 @@
+	@Operation(summary = "지원서 캘린더 월간 데이터 조회")
+	@ApiResponses(
+		value = {@ApiResponse(responseCode = "200", description = "캘린더 월간 데이터 조회 성공"),
+			@ApiResponse(responseCode = "400", description = "날짜 정보가 잘못되었음")
```
Swagger에 errorCode가 명시되어 있지 않은 상황인데요!
[PR](https://github.com/team-go-higher/backend/pull/56) 보시면 Swagger의 ExampleObject로 표현할 수 있는 방법이 있습니다! 보고 반영하시면 좋을 것 같아요!

Seongwon97 replied:
PR확인하여 추가했습니다!!

Byeongju-Kong commented on out-adapter-persistence-jpa/src/main/java/gohigher/application/entity/ApplicationJpaEntity.java:
```diff
@@ -75,12 +76,31 @@ public static ApplicationJpaEntity of(Long userId, Application application) {
+		processes.add(process);
+	}
+
+	public Application convertToDomain() {
```
toDomain 정도로 해도 되지 않을까라는 생각이 듭니다!

Seongwon97 replied:
`toDomain()`으로 네이밍 수정했습니다!

Byeongju-Kong commented on out-adapter-persistence-jpa/src/main/java/gohigher/application/entity/ApplicationJpaEntity.java:
```diff
@@ -75,12 +76,31 @@ public static ApplicationJpaEntity of(Long userId, Application application) {
+
+	public Application convertToDomain() {
+		List<Process> processes = this.processes.stream()
+			.sorted(Comparator.comparingInt(ApplicationProcessJpaEntity::getOrder))
```
Process가 Comparable을 구현하는 방식은 고려해보셨을까요? 해보셨다면 그렇게 하지 않으신 이유가 궁금합니다!

Seongwon97 replied:
해당 부분은 고민중에 있었습니다.
Core의 domain이나 usecase 모듈에서 로직적으로 order정렬을 할 일이 있다면 Process에 `Comparable`을 구현하는 것이 훨씬 효율적이라 생각합니다.
하지만 현재의 로직은 Core의 로직을 만지는 것이 아닌 외부 out-adapter에서 domain으로 변환하는 로직이라, out-adapter의 로직의 편의를 위해 Domain에 추가 구현을 한다는 느낌이 들어 보류하게 되었습니다. (살짝 main코드에 test의 편의를 위한 로직을 추가하는..?그런 비슷한 느낌이었습니다.)

만약 core의 비즈니스 로직에서 Process의 정렬 편의를 위해 Comparable을 구현한다면 그때 해당 코드도 추가로 변경하면 좋지 않을까 합니다.

Byeongju-Kong reviewed (changes requested):
고생하셨습니다~! 몇가지 리뷰 남겨두었어요!
유빈님과 성원님 두분 다 BeforeEach에서 테스트 하려는 객체를 생성해서 초기화하시던데, 혹시 특별한 이유가 있을까요? 저는 테스트 하려는 객체가 상태 값을 가지지 않는다면 필드 초기화로 충분하다고 생각하는데요! 의견 달아주시면 감사하겠습니다~!

Byeongju-Kong commented on core-application/build.gradle:
```diff
@@ -16,6 +16,9 @@ dependencies {
     implementation 'jakarta.validation:jakarta.validation-api'
+
+    testImplementation(testFixtures(project(":core-domain")))
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
core-application 모듈은 mockito를 통해서 테스트 하는데 SpringBootTest 의존성이 필요한가요?

Seongwon97 replied:
저의 편의를 위해..`spring-boot-starter-test`의존성을 추가한 것 같습니다..
아래와 같이 상황별 적합한 의존성을 추가하도록 변경했습니다!
```
    testImplementation 'org.mockito:mockito-core:5.3.1'
    testImplementation 'org.mockito:mockito-junit-jupiter:5.3.1'
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.3'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.3'
    testImplementation "org.assertj:assertj-core:3.24.2"
```

Byeongju-Kong reviewed (commented):
> 응답 값에서 data 내부에 applications으로 감싸고 안에 List형태로 반환해줄지 data안에 바로 List를 반환할지 고민입니다.

data: {
}

에서 data라는 key 값을 제외하고 생각한다면, 원래의 포맷은 {} 안에 그냥 바디가 들어가도 되는 것 같습니다!
그래서 그냥 data 안에 List를 응답해도 되지 않을까요?

> TestFixture를 원래 domain모듈에 위치시키고 다른 모듈들에서 각각 필요한 형태로 변환한 후에 사용하려했는데...모듈간 implementation을 하여도 main패키지만 의존이 넘어가고 test패키지는 넘어가지 않는다네요ㅠㅠ 이렇게 되면 TestFixture를 모듈간 따로 관리해야하는데..이것 또한 너무 많은 비용이 들 것 같아 고민입니다.

도메인에 TestFixture를 정의해두고 사용하면 될 것 같은데요! 제가 정확히 성원님이 고민하시는 부분을 이해하지 못했어요😂, 혹시나 해서 제가 멀티 모듈 전환시에 테스트 픽스처를 구성할 때 참고했던 글인데 참고하시면 성원님의 고민이 해결될 수 도 있을 것 같아서 첨부합니다!
 [글1](https://www.google.com/search?q=%EB%A9%80%ED%8B%B0%EB%AA%A8%EB%93%88+testFixture&rlz=1C5CHFA_enKR959KR959&oq=%EB%A9%80%ED%8B%B0%EB%AA%A8%EB%93%88+testFixture&aqs=chrome..69i57j33i160.5162j0j7&sourceid=chrome&ie=UTF-8), [글2](https://livenow14.tistory.com/79)

Byeongju-Kong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/entity/ApplicationRepositoryTest.java:
```diff
@@ -0,0 +1,155 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
+
+@DisplayName("ApplicationRepository 클래스의")
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@DataJpaTest
+class ApplicationRepositoryTest {
```
지금 ApplicationPersistenceQueryAdapterTest도 있고 ApplicationRepository 테스트도 있는데요!

ApplicationPersistenceQueryAdapterTest를 mock test로 진행하지 않고 실제 DB로 진행하고
ApplicationRespository 테스트는 ApplicationPersistenceQueryAdapterTest를 하면서 간접적으로 하는건 어떨까요?

Seongwon97 replied:
저는 두 테스트의 책임이 좀 다르다 생각하여 분리했습니다.
Repository는 하나의 메서드당 데이터베이스에 하나의 쿼리를 실행하여 CRUD작업을 하는 것이 관심사이며 Adapter의 경우, core에서 원하는 작업(CRUD)을 진행하기 위해 1개 이상의 쿼리를 실해하여 원하는 응답값을 반환하는 것이 관심사라 생각합니다.
즉, Adapter는 자신의 작업을 하기 위해 여러 Repository 메서드를 호출하여 작업을 수행하고 core에서 정의한 명세로 Entity를 Domain으로 변환해주는 2가지 책임을 가진다 생각하고 있습니다. 

그리하여 내가 만든 쿼리가 올바르게 동작하는지에 대한 체크는 Repository테스트로 단위테스트를 수행하는게 더 좋을 것이라 생각됩니다.
대신 Repository에 대한 테스트는 기본적으로 DataJpa에서 제공하는 쿼리가 아닌 우리가 커스텀하여 JPQL문을 만들거나 QueryDsl을 통해 동적 쿼리를 만드는 것에 대해서만 테스트하면 충분할 것 같습니다!

kyukong replied:
저도 성원님 의견에 동의합니다!

Repository 테스트 관해서는 전 개인적으로 '정의된' 모든 메서드를 테스트하는 편입니다. (네이밍으로 쿼리 생성되는 메서드 포함) 네이밍 규칙이 있고 그에 따라 쿼리문이 생성되므로, 해당 규칙을 잘 지켰는지에 대한 안정을 위해 테스트하는 편입니다.

Byeongju-Kong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/entity/ApplicationRepositoryTest.java:
```diff
@@ -0,0 +1,155 @@
+			}
+
+			@Test
+			@DisplayName("2개성의 과정을 모두 담은 하나의 지원 공고를 반환한다.")
```
오타가 있는 것 같아요!

Seongwon97 replied:
감사합니다! 오타 수정했습니다!

Byeongju-Kong commented on out-adapter-persistence-jpa/src/main/java/gohigher/application/ApplicationPersistenceQueryAdapter.java:
```diff
@@ -0,0 +1,27 @@
+package gohigher.application;
+
+import java.util.List;
+
+import org.springframework.stereotype.Component;
+
+import gohigher.application.entity.ApplicationJpaEntity;
+import gohigher.application.entity.ApplicationRepository;
+import gohigher.application.port.out.persistence.ApplicationPersistenceQueryPort;
+import lombok.RequiredArgsConstructor;
+
+@Component
+@RequiredArgsConstructor
+public class ApplicationPersistenceQueryAdapter implements ApplicationPersistenceQueryPort {
+
+	private final ApplicationRepository applicationRepository;
+
+	@Override
+	public List<Application> findByIdAndMonth(Long userId, int year, int month) {
+		List<ApplicationJpaEntity> applicationJpaEntities =
+			applicationRepository.findByUserIdAndDate(userId, year, month);
+
+		return applicationJpaEntities.stream()
+			.map(ApplicationJpaEntity::convertToDomain)
+			.toList();
+	}
+}
```
도메인 단위 package가 분리 되지 않았네요!

Seongwon97 replied:
혹시 어떤 패키지 분리를 말씀하신걸까요?? 현재 main 클래스와 똑같이 `package gohigher.application;` 경로에 위치해있습니다!

Seongwon97 committed:
refactor: ApplicationJpaEntity의 도메인 변환 메서드 네이밍 수정

- convertToDomain -> toDomain

Seongwon97 committed:
feat: ApplicationProcessJpaEntity의 도메인 변환 메서드 생성

Seongwon97 committed:
build: 불필요한 spring-boot-starter-test 의존성을 제거하고 필요한 의존성 추가

Seongwon97 committed:
feat: year, month의 유효 범위를 검증하는 Request객체 생성

Seongwon97 committed:
refactor: controller과 service레이어의 통신을 Request객체를 통해 하도록 수정

Seongwon97 committed:
feat: swagger문서에 실해했을 때의 메시지 정보 추가

Seongwon97 committed:
build: application 모듈에 junit params 의존성 추가

Seongwon97 committed:
test: 잘못된 날짜 정보를 요청을 검증하는 테스트 추가

Seongwon97 committed:
test: 잘못된 DisplayName 수정

Seongwon97 committed:
test: repository테스트에서 db에 존재하는 데이터로 인한 테스트 격리를 위해 userId 수정

Seongwon97 committed:
refactor: 캘린더 조회의 반환값을 data하위로 바로 list를 반환하도록 수정

Seongwon97 commented:
@Byeongju-Kong 
> `data: {}` 에서 data라는 key 값을 제외하고 생각한다면, 원래의 포맷은 {} 안에 그냥 바디가 들어가도 되는 것 같습니다!
> 그래서 그냥 data 안에 List를 응답해도 되지 않을까요?
- 해당 부분도 depth를 줄이고 data에서 바로 list를 반환하도록 수정했습니다.

> 도메인에 TestFixture를 정의해두고 사용하면 될 것 같은데요! 제가 정확히 성원님이 고민하시는 부분을 이해하지 못했어요😂, 혹시나 해서 제가 멀티 모듈 전환시에 테스트 픽스처를 구성할 때 참고했던 글인데 참고하시면 성원님의 고민이 해결될 수 도 있을 것 같아서 첨부합니다!
- 해당 부분도 앞서 `java-test-fixtures`를 통해 해결했습니다!! 위의 글을 줄이는 것을 깜빡했네요😭

kyukong commented on core-application/build.gradle:
```diff
@@ -16,6 +16,14 @@ dependencies {
     implementation 'com.fasterxml.jackson.core:jackson-annotations:2.15.2'
     implementation 'jakarta.validation:jakarta.validation-api'
+
+    testImplementation(testFixtures(project(":core-domain")))
+    testImplementation 'org.mockito:mockito-core:5.3.1'
+    testImplementation 'org.mockito:mockito-junit-jupiter:5.3.1'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.3'
+    testImplementation 'org.junit.jupiter:junit-jupiter-params:5.9.3'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.3'
+    testImplementation "org.assertj:assertj-core:3.24.2"
```
```suggestion
    testImplementation(testFixtures(project(":core-domain")))
    
    testImplementation 'org.mockito:mockito-core:5.3.1'
    testImplementation 'org.mockito:mockito-junit-jupiter:5.3.1'
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.3'
    testImplementation 'org.junit.jupiter:junit-jupiter-params:5.9.3'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.3'
    testImplementation "org.assertj:assertj-core:3.24.2"
```

ㅎㅎ.. 위와 동일하게 가면 어떨까요..

Seongwon97 replied:
공백라인 추가했습니다!!!

kyukong commented on core-application/src/main/java/gohigher/application/port/in/CalenderApplicationRequest.java:
```diff
@@ -0,0 +1,34 @@
+		this.month = month;
+	}
+
+	private void validateNegativeYear(int year) {
+		if (year <= 0) {
+			throw new GoHigherException(INVALID_DATE_INFO);
+		}
+	}
+
+	private void verifyMonthIsInValidRange(int month) {
+		if (month < 1 || month > 12) {
+			throw new GoHigherException(INVALID_DATE_INFO);
+		}
+	}
```
```suggestion
	private void validateNegativeYear(int year) {
		if (year <= 0) {
			throw new GoHigherException(INVALID_DATE_INFO);
		}
	}

	private void validateMonthIsWithinRange(int month) {
		if (month < 1 || month > 12) {
			throw new GoHigherException(INVALID_DATE_INFO);
		}
	}
```

해당 검증 로직은 DTO 가 아닌 도메인이 어울릴듯 합니다! 그리고 검증 테스트 관련해서는 용어(validate) 를 통일하면 어떨까요!

Seongwon97 replied:
해당 메서드도 validate~로 수정했습니다!!

kyukong commented on core-application/src/main/java/gohigher/application/service/ApplicationQueryService.java:
```diff
@@ -0,0 +1,44 @@
+		return calenderApplicationResponses;
+	}
+
+	private void addScheduleResponses(Application application,
+		List<CalenderApplicationResponse> calenderApplicationRespons) {
+		for (Process process : application.getProcesses()) {
+			calenderApplicationRespons.add(
+				new CalenderApplicationResponse(application.getCompanyName(), process.getType().name(),
+					process.getSchedule()));
+		}
+	}
```
```suggestion
	private void addScheduleResponse(Application application,
		List<CalenderApplicationResponse> calenderApplicationResponse) {
		for (Process process : application.getProcesses()) {
			calenderApplicationRespons.add(
				new CalenderApplicationResponse(application.getCompanyName(), process.getType().name(),
					process.getSchedule()));
		}
	}
```

Domain -> DTO 변환로직은 DTO 에 두면 어떨까요? 제가 좋아하는 방법입니다. ㅎㅎ 해당 로직을 서비스에 두면 서비스에 검증로직만 많이 차지하여 가독성이 떨이지더라구요!

Seongwon97 replied:
해당 로직은 여러 지원서들이 갖는 프로세스를 각각의 DTO로 변환하는 코드라 DTO로 옮길 수가 없습니다ㅠ
만약 response를 `List<CalenderApplicationResponse>`를 바로 반환하는 것이 아닌 해당 값을 필드로 갖는 새로운 DTO로 감싸서 반환한다면 새로운 DTO에 위의 로직을 넣을 수 있을 것 같은데, 현재 응답값은 List로 바로 반환하는 것이라 위치를 옮기는 것은 힘들것 같습니다ㅠㅠ

kyukong commented on core-application/src/test/java/gohigher/application/port/in/CalenderApplicationRequestTest.java:
```diff
@@ -0,0 +1,55 @@
+			@ParameterizedTest
+			@ValueSource(ints = {0, -1})
+			void it_throw_exception(int year) {
+				GoHigherException goHigherException = assertThrows(GoHigherException.class,
+					() -> new CalenderApplicationRequest(userId, year, month));
+				assertThat(goHigherException.getErrorCode()).isEqualTo(INVALID_DATE_INFO.getErrorCode());
```
`assertThatThrownBy` 를 사용하지 않은건 에러코드도 검증하기 위함인가요?? 혹시 이 테스트에서 에러가 발생한다면 assertThrows 에서 발생하는지, assertThat 에서 발생하는지 구분이 가는지 궁금합니다!

Seongwon97 replied:
맞습니다. 초반에는 DisplayName자체를 ~~ ErrorCode를 반환한다는 이름으로 두어 ErrorCode 올바르게 반환하는지에 초점을 잡고 테스트를 했습니다.
하지만 지금은 DisplayName자체를 enum상수 값에 대한 검증으로 변경하였기에 굳이 exception인스턴스를 뽑지 않아도 될 것 같아서 assertThatThrownBy로 변경했습니다.

kyukong commented on core-domain/src/testFixtures/java/gohigher/application/ApplicationFixture.java:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	@NoArgsConstructor
+	public static class Builder {
```
롬복 빌더 어노테이션이 아닌 직접 구현한 이유가 궁금합니다!

Seongwon97 replied:
롬복의 `@Builder`의 경우 class타입에만 적용할 수 있습니다ㅠ그래서 enum타입에는 불가합니다

kyukong commented on in-adapter-api/src/main/java/gohigher/application/ApplicationQueryController.java:
```diff
@@ -0,0 +1,30 @@
+	@GetMapping("/v1/applications/calender")
+	public ResponseEntity<GohigherResponse<List<CalenderApplicationResponse>>> findByMonth(@Login Long userId,
+		@RequestParam int year, @RequestParam int month) {
+		CalenderApplicationRequest request = new CalenderApplicationRequest(userId, year, month);
```
`@ModelAttribute` 를 사용하는건 어떨까요?

Seongwon97 replied:
아직 @Valid 에러를 잡는 체계가 확실히 구축되지 않아 현재는 dto변환 과정에서 직접 예외를 날려줬는데 `@ModelAttribute`를 사용하고 dto의 검증 로직도 `@Valid`를 사용할지 고민이긴 합니다😭

kyukong commented on out-adapter-persistence-jpa/src/test/java/gohigher/application/ApplicationPersistenceQueryAdapterTest.java:
```diff
@@ -0,0 +1,96 @@
+					.orElseThrow();
+
+				assertAll(
+					() -> assertThat(response).hasSize(2), // naverApplicatio, kakaoApplication
```
```suggestion
					() -> assertThat(response).hasSize(2), // naverApplication, kakaoApplication
```

kyukong reviewed (approved):
안녕하세요 성원님!
간단한 리뷰들이라 바로 approve 하겠습니다. ㅎㅎ
API 구현 초반이라 아직 결정해야 하는 것들이 많았을텐데 고생하셨습니다!

Seongwon97 committed:
build: 의존성 사이 공백라인 추가

Seongwon97 committed:
refactor: 올바른 월인지 검증하는 로직 메서드 명 수정

Seongwon97 committed:
test: 주석 오타 수정

Seongwon97 committed:
Merge branch 'develop' into feature/#48/schedule-query-api

# Conflicts:
#	core-application/build.gradle
#	core-domain/build.gradle
#	out-adapter-persistence-jpa/src/test/java/gohigher/GoHigherApplicationTest.java

Seongwon97 committed:
build: 머지 과정에서 빠진 java-test-fixtures 의존 추가

Seongwon97 committed:
test: 예외 검증 메서드를 assertThatThrownBy으로 변경

Seongwon97 committed:
fix: 잘못된 import경로 수정

shindong96 commented on core-application/src/main/java/gohigher/application/service/ApplicationQueryService.java:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	private void addScheduleResponses(Application application,
+		List<CalenderApplicationResponse> calenderApplicationRespons) {
```
```suggestion
		List<CalenderApplicationResponse> calenderApplicationResponse) {
```
오타 발견!!ㅋㅋ

Seongwon97 replied:
수정했습니다!

shindong96 commented on core-application/src/main/java/gohigher/application/service/ApplicationQueryService.java:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	private void addScheduleResponses(Application application,
+		List<CalenderApplicationResponse> calenderApplicationRespons) {
```
```suggestion
	private List<CalenderApplicationResponse> addScheduleResponses(Application application) {
```
이렇게 List를 반환하는 방법은 어떨까요? 자세히 기억이 안나는데 빈 배열을 파라미터로 넣어주고 외부에서 값을 채우는 방식의 단점이 있다고 들었습니다. 이유는 다시 한 번 찾아볼게요ㅋㅋ

Seongwon97 replied:
로직들을 stream으로 처리하도록 변경했습니다! 
별도로 이야기 나눈대로 map관련해서만 분리했습니다

shindong96 commented on core-application/src/main/java/gohigher/application/service/ApplicationQueryService.java:
```diff
@@ -0,0 +1,44 @@
+
+		List<CalenderApplicationResponse> calenderApplicationResponses = new ArrayList<>();
+		for (Application application : applications) {
+			addScheduleResponses(application, calenderApplicationResponses);
```
개인적인 사견입니다! 현재 Service 메서드의 로직에 Response을 만드는 부분이 큰 비중을 차지하는 것 같은데요.
개인적으로는 `CalendarApplicationResponses` 객체를 만드는게 어떨까요? 이 객체를 만들어서 `List<CalendarApplicationResponse>`를 field로 가지게 한다면 Responses 객체에 정적 팩토리 메서드를 사용해서 응답할 수 있을 거 같아요! 그렇다면 서비스 로직이 좀 더 비즈니스 로직만을 남겨둘 수 있을 거 같습니다!

한가지 의문이 있다면 해당 Response는 사실 `간단 지원 카드`를 return 해주는 것으로 이해하고 있습니다! 때문에 네이밍을 `SimpleApplicaionResponse`와 같이 변경하는 건 어떨까요?

Seongwon97 replied:
흠....초반에 그런 구조였으나 [병주님의 코멘트](https://github.com/team-go-higher/backend/pull/52#pullrequestreview-1617536154) 를 반영해서 depth를 한 단게 없애고 `List<CalendarApplicationResponse>`를 반환하는 구조로 변경했었습니다...!

Seongwon97 replied:
`SimpleApplicaionResponse`의 의미를 주더라도 Calender의 네이밍은 줘야 한다고 생각합니다.
지원서 간편 조회 응답이 칸반에도 존재하고 캘린더에도 존재하는데, 단순히 `SimpleApplicaionResponse`이라는 네이밍을 주면 둘 사이의 혼동이 올 것 같다는 생각이 들어요.

둘을 같이 사용하기에는 칸반과 캘린더가 원하는 데이터들이 서로 다르다는 점이 있기에 분리하는 것이 좋다 생각하고요!

shindong96 commented on core-application/src/main/java/gohigher/application/port/in/CalenderApplicationResponse.java:
```diff
@@ -0,0 +1,15 @@
+
+@Getter
+@RequiredArgsConstructor
+public class CalenderApplicationResponse {
```
해당 Response가 return해야하는 정보들에 대해 따로 이야기 나눠보면 좋을 거 같아요!
일단 그래도 Application과 Process의 PK는 넘겨줘야 수정이나 조회에 유용할 거 같아요 :)

Seongwon97 replied:
PK 값 추가했습니다!!

shindong96 commented on in-adapter-api/src/main/java/gohigher/application/ApplicationQueryController.java:
```diff
@@ -0,0 +1,30 @@
+
+	@GetMapping("/v1/applications/calender")
+	public ResponseEntity<GohigherResponse<List<CalenderApplicationResponse>>> findByMonth(@Login Long userId,
+		@RequestParam int year, @RequestParam int month) {
```
그냥 궁금증입니다!
혹시 Body가 아닌 Param으로 받으신 이유가 있을까요? 필터링의 관점이기 때문인가요??

Seongwon97 replied:
음 일단 GET의 body를 넣는 것은 논쟁거리중 하나입니다.
GET의 body에 대해 간단히 설명드리자면 과거에 GET Method에 body를 넣는 것은 허용되지 않았습니다. 하지만 2014년부터 HTTP 스펙에서 공식적으로 지원하기로 한 것으로 알고 있습니다.

하지만 과거에 지원을 하지 않았던 아래의 RFC7231문서를 보면 요청이 거부될 수 있는 문제가 발생할 수 있다는 내용이 있긴하여서..저는 GET 요청의 경우 body가 아닌 queryParam으로 보내는 편입니다.

![image](https://github.com/team-go-higher/backend/assets/57744251/afce54f9-d903-49df-ad7a-88311ae713d3)

shindong96 replied:
아! GET이라서군요!! 이해했습니다!!

shindong96 commented on out-adapter-persistence-jpa/src/main/java/gohigher/application/entity/ApplicationRepository.java:
```diff
@@ -1,6 +1,16 @@
+		+ "JOIN FETCH a.processes p "
+		+ "WHERE a.userId = :userId "
+		+ "AND FUNCTION('YEAR', p.schedule) = :year "
+		+ "AND FUNCTION('MONTH', p.schedule) = :month")
```
혹시 해당 쿼리의 FUNCTION 부분 설명 부탁드려도 될까요?? 😅

Seongwon97 replied:
FUNCTION('MONTH', p.schedule): 이 부분은 JPQL에서 사용하는 특수 함수를 호출하는 부분입니다. 여기서 'MONTH' 함수는 주어진 날짜나 시간 값에서 월 (Month) 정보를 추출하는 함수를 호출하는 것을 나타냅니다.

shindong96 replied:
🫢

shindong96 reviewed (changes requested):
성원님 리뷰가 늦어 죄송합니다!!
함께 의견 나눴으면 좋겠는 부분을 적었습니다
한번  보시구 코멘트 부탁드려요!

Seongwon97 committed:
feat: core-domain 지원서 정보와 지원서 조회 응답값에 id정보 추가

Seongwon97 committed:
fix: 테스트 의존성 충돌 해결

Seongwon97 committed:
Merge remote-tracking branch 'origin/develop' into feature/#48/schedule-query-api

# Conflicts:
#	core-domain/src/main/java/gohigher/application/Application.java
#	core-domain/src/main/java/gohigher/common/JobInfo.java
#	core-domain/src/main/java/gohigher/recruitment/Recruitment.java
#	in-adapter-api/src/main/java/gohigher/application/ApplicationCommandController.java
#	out-adapter-persistence-jpa/build.gradle

Seongwon97 committed:
refactor: 필드명 오타 수정

Seongwon97 committed:
refactor: 일정 조회 응답 객체 생성 로직 수정

Seongwon97 committed:
feat: 도메인과 캘린더 조회 response에 processId 추가

shindong96 reviewed (approved):
Conversation이 어마무시 하네요...
너무 고생하셨습니다 😂

Seongwon97 committed:
Merge remote-tracking branch 'origin/develop' into feature/#48/schedule-query-api

# Conflicts:
#	core-domain/build.gradle
